rp2040-hal/src/spi.rs

```diff
@@ -1,3 +1,214 @@
 //! Serial Peripheral Interface (SPI)
-// See [Chapter 4 Section 4](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
-// TODO
+//!
+//! See [Chapter 4 Section 4](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+//!
+//! ## Usage
+//!
+//! ```rust
+//! let sio = Sio::new(p.SIO);
+//! let gpio_pins = Pins::new(p.IO_BANK0, p.PADS_BANK0, sio.gpio_bank0, &mut p.RESETS);
+//!
+//! let _ = gpio_pins.spi_sclk.into_mode::<FunctionSpi>();
+//! let _ = gpio_pins.spi_mosi.into_mode::<FunctionSpi>();
+//!
+//! let spi = Spi::<_, _, 8>::new(p.SPI0).init(&mut p.RESETS, 125_000_000, 16_000_000, &MODE_0);
+//! ```
+
+use crate::resets::SubsystemReset;
+use core::{convert::Infallible, marker::PhantomData, ops::Deref};
+use embedded_hal::blocking::spi;
+use embedded_hal::spi::{FullDuplex, Mode, Phase, Polarity};
+use pac::RESETS;
+
+/// State of the SPI
+pub trait State {}
+
+/// Spi is disabled
+pub struct Disabled {
+    __private: (),
+}
+
+/// Spi is enabled
+pub struct Enabled {
+    __private: (),
+}
+
+impl State for Disabled {}
+impl State for Enabled {}
+
+/// Pac SPI device
+pub trait SpiDevice: Deref<Target = pac::spi0::RegisterBlock> + SubsystemReset {}
+
+impl SpiDevice for pac::SPI0 {}
+impl SpiDevice for pac::SPI1 {}
+
+/// Data size used in spi
+pub trait DataSize {}
+
+impl DataSize for u8 {}
+impl DataSize for u16 {}
+
+/// Spi
+pub struct Spi<S: State, D: SpiDevice, const DS: u8> {
+    device: D,
+    state: PhantomData<S>,
+}
+
+impl<S: State, D: SpiDevice, const DS: u8> Spi<S, D, DS> {
+    fn transition<To: State>(self, _: To) -> Spi<To, D, DS> {
+        Spi {
+            device: self.device,
+            state: PhantomData,
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> D {
+        self.device
+    }
+}
+
+impl<D: SpiDevice, const DS: u8> Spi<Disabled, D, DS> {
+    /// Create new spi device
+    pub fn new(device: D) -> Spi<Disabled, D, DS> {
+        Spi {
+            device,
+            state: PhantomData,
+        }
+    }
+    /// Set baudrate based on peripheral clock
+    ///
+    /// Typically the peripheral clock is set to 125_000_000
+    fn set_baudrate(&mut self, peri_frequency: u32, baudrate: u32) -> u32 {
```

Comment:
Should I use embedded_time::rate::*Hertz here for the frequency?

Reply:
That would make this easier to use, at least.